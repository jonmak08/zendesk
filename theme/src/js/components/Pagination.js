import preact from 'preact';
import PropTypes from 'prop-types';

import times from 'lodash.times';

const PAGE_BUFFER = 5;

class PaginationItem extends preact.Component {
	constructor(props) {
		super(props);

		this.handleClick = this.handleClick.bind(this);
	}

	handleClick() {
		const {number, onClick} = this.props;

		onClick(number);
	}

	render({active, label, number}) {
		return (
			<li class={active ? 'pagination-current' : ''}>
				{!active && (
					<button aria-label={`go to page ${label}`} class="btn-unstyled" onClick={this.handleClick} type="button" value={number}>
						{label}
					</button>
				)}

				{active && (
					<span aria-current="true" aria-label={`current page, page ${label}`}>{label}</span>
				)}
			</li>
		);
	}
}

PaginationItem.PropTypes = {
	active: PropTypes.bool,
	label: PropTypes.oneOfType(
		[PropTypes.number, PropTypes.string]
	),
	number: PropTypes.number,
	onClick: PropTypes.func.isRequired
};

class Pagination extends preact.Component {
	constructor(props) {
		super(props);

		this.getPages = this.getPages.bind(this);
		this.handleClick = this.handleClick.bind(this);
		this.setBuffers = this.setBuffers.bind(this);

		this.state = {
			currentPage: 1
		};
	}

	getPages() {
		const {total} = this.props;
		const {currentPage} = this.state;

		const {nextPageBuffer, prevPageBuffer} = this.setBuffers();

		const startPage = currentPage - prevPageBuffer;
		const totalPages = nextPageBuffer + prevPageBuffer + 1;

		const pages = times(
			totalPages,
			i => {
				const pageNumber = startPage + i;

				return {
					label: pageNumber,
					number: pageNumber
				};
			}
		);

		if (nextPageBuffer) {
			pages.push(
				{
					label: '›',
					number: (currentPage + 1)
				},
				{
					label: '»',
					number: total
				}
			);
		}

		if (prevPageBuffer) {
			pages.unshift(
				{
					label: '«',
					number: 1
				},
				{
					label: '‹',
					number: (currentPage - 1)
				}
			);
		}

		return pages;
	}

	handleClick(activePage) {
		const {onClick} = this.props;

		this.setState(
			{
				currentPage: activePage
			}
		);

		onClick(activePage);
	}

	setBuffers() {
		const {total} = this.props;
		const {currentPage} = this.state;

		let nextPageBuffer = PAGE_BUFFER;
		let prevPageBuffer = 0;

		if (currentPage > 1) {
			prevPageBuffer = currentPage > PAGE_BUFFER ? PAGE_BUFFER : currentPage - 1;

			if (currentPage === total) {
				nextPageBuffer = 0;
			}
		}
		else {
			nextPageBuffer = total - currentPage >= PAGE_BUFFER ? PAGE_BUFFER : total - currentPage;
		}

		return {
			nextPageBuffer,
			prevPageBuffer
		};
	}

	render() {
		return (
			<nav aria-label="pagination navigation" class="pagination" role="navigation">
				<ul>
					{this.getPages().map(
						page => (
							<PaginationItem
								active={this.state.currentPage === page.number}
								label={page.label}
								number={page.number}
								onClick={this.handleClick}
							/>
						)
					)}
				</ul>
			</nav>
		);
	}
}

Pagination.PropTypes = {
	onClick: PropTypes.func.isRequired,
	total: PropTypes.number.isRequired
};

export default Pagination;