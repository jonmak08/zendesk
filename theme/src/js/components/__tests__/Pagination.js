import preact from 'preact';
import {
	cleanup,
	debounceRenderingOff,
	fireEvent,
	render
} from 'preact-testing-library';

import Pagination from '../Pagination';

const setup = () => {
	const handleClick = jest.fn();

	const utils = render(
		<Pagination onClick={handleClick} perPage={10} total={100} />
	);

	const firstPage = utils.queryByText('«');
	const lastPage = utils.queryByText('»');
	const nextPage = utils.queryByText('›');
	const prevPage = utils.queryByText('‹');

	return {
		firstPage,
		lastPage,
		nextPage,
		prevPage,
		...utils
	};
};

afterEach(cleanup);

describe('Pagination', () => {
	it('renders correctly', () => {
		const {container} = setup();

		expect(container).toMatchSnapshot();
	});

	it('renders page one as the first page on initial load', () => {
		const {getByText} = setup();

		const page1 = getByText('1');

		expect(page1).toBeTruthy();
	});

	it('renders only next and last page controls on initial load', () => {
		const {firstPage, lastPage, nextPage, prevPage} = setup();

		expect(firstPage).toBeNull();
		expect(lastPage).toBeTruthy();
		expect(nextPage).toBeTruthy();
		expect(prevPage).toBeNull();
	});

	it('renders last page correctly', () => {
		debounceRenderingOff();

		const {lastPage, getByText} = setup();

		fireEvent.click(lastPage);

		const page100 = getByText('100');

		expect(page100).toBeTruthy();
	});

	it('renders only first and prev page controls on last page', () => {
		debounceRenderingOff();

		const {lastPage, queryByText} = setup();

		fireEvent.click(lastPage);

		const newFirstPage = queryByText('«');
		const newLastPage = queryByText('»');
		const newNextPage = queryByText('›');
		const newPrevPage = queryByText('‹');

		expect(newFirstPage).toBeTruthy();
		expect(newLastPage).toBeNull();
		expect(newNextPage).toBeNull();
		expect(newPrevPage).toBeTruthy();
	});

	it('renders all page controls when paginate on a page that is neither the first nor last page', () => {
		debounceRenderingOff();

		const {queryByText} = setup();

		fireEvent.click(queryByText('2'));

		const newFirstPage = queryByText('«');
		const newLastPage = queryByText('»');
		const newNextPage = queryByText('›');
		const newPrevPage = queryByText('‹');

		expect(newFirstPage).toBeTruthy();
		expect(newLastPage).toBeTruthy();
		expect(newNextPage).toBeTruthy();
		expect(newPrevPage).toBeTruthy();
	});

	it('renders a maximum of five pages previous to and after the current page', () => {
		debounceRenderingOff();

		const {container, queryByText} = setup();

		fireEvent.click(queryByText('6'));

		expect(container).toMatchSnapshot();
	});
});