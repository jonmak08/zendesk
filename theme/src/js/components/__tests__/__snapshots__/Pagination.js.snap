// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders a maximum of five pages previous to and after the current page 1`] = `
<div>
  <nav
    aria-label="pagination navigation"
    class="pagination"
    role="navigation"
  >
    <ul>
      <li
        class=""
      >
        <button
          aria-label="go to page «"
          class="btn-unstyled"
          type="button"
          value="1"
        >
          «
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page ‹"
          class="btn-unstyled"
          type="button"
          value="5"
        >
          ‹
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 1"
          class="btn-unstyled"
          type="button"
          value="1"
        >
          1
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 2"
          class="btn-unstyled"
          type="button"
          value="2"
        >
          2
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 3"
          class="btn-unstyled"
          type="button"
          value="3"
        >
          3
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 4"
          class="btn-unstyled"
          type="button"
          value="4"
        >
          4
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 5"
          class="btn-unstyled"
          type="button"
          value="5"
        >
          5
        </button>
        
      </li>
      <li
        class="pagination-current"
      >
        
        <span
          aria-current="true"
          aria-label="current page, page 6"
        >
          6
        </span>
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 7"
          class="btn-unstyled"
          type="button"
          value="7"
        >
          7
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 8"
          class="btn-unstyled"
          type="button"
          value="8"
        >
          8
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 9"
          class="btn-unstyled"
          type="button"
          value="9"
        >
          9
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 10"
          class="btn-unstyled"
          type="button"
          value="10"
        >
          10
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 11"
          class="btn-unstyled"
          type="button"
          value="11"
        >
          11
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page ›"
          class="btn-unstyled"
          type="button"
          value="7"
        >
          ›
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page »"
          class="btn-unstyled"
          type="button"
          value="100"
        >
          »
        </button>
        
      </li>
    </ul>
  </nav>
</div>
`;

exports[`Pagination renders correctly 1`] = `
<div>
  <nav
    aria-label="pagination navigation"
    class="pagination"
    role="navigation"
  >
    <ul>
      <li
        class="pagination-current"
      >
        
        <span
          aria-current="true"
          aria-label="current page, page 1"
        >
          1
        </span>
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 2"
          class="btn-unstyled"
          type="button"
          value="2"
        >
          2
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 3"
          class="btn-unstyled"
          type="button"
          value="3"
        >
          3
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 4"
          class="btn-unstyled"
          type="button"
          value="4"
        >
          4
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 5"
          class="btn-unstyled"
          type="button"
          value="5"
        >
          5
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page 6"
          class="btn-unstyled"
          type="button"
          value="6"
        >
          6
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page ›"
          class="btn-unstyled"
          type="button"
          value="2"
        >
          ›
        </button>
        
      </li>
      <li
        class=""
      >
        <button
          aria-label="go to page »"
          class="btn-unstyled"
          type="button"
          value="100"
        >
          »
        </button>
        
      </li>
    </ul>
  </nav>
</div>
`;
